lib/Doctrine/ORM/Persisters/ManyToManyPersister.php

```diff
@@ -37,130 +37,288 @@ class ManyToManyPersister extends AbstractCollectionPersister
+        $params         = array();
+        $mapping        = $coll->getMapping();
+        $association    = $mapping;
+        $class          = $this->em->getClassMetadata($mapping['sourceEntity']);
+        $id             = $this->uow->getEntityIdentifier($coll->getOwner());
+
+        if ( ! $mapping['isOwningSide']) {
+            $targetEntity   = $this->em->getClassMetadata($mapping['targetEntity']);
+            $association    = $targetEntity->associationMappings[$mapping['mappedBy']];
+        }
+
+        $joinColumns = ( ! $mapping['isOwningSide'])
+            ? $association['joinTable']['inverseJoinColumns']
+            : $association['joinTable']['joinColumns'];
+
+        foreach ($joinColumns as $joinColumn) {
+            $columnName     = $this->quoteStrategy->getJoinColumnName($joinColumn, $class, $this->platform);
+            $referencedName = $joinColumn['referencedColumnName'];
+            $conditions[]   = 't.' . $columnName . ' = ?';
+            $params[]       = ($class->containsForeignIdentifier)
+                ? $id[$class->getFieldForColumn($referencedName)]
+                : $id[$class->fieldNames[$referencedName]];
+        }
+
+        $joinTableName = $this->quoteStrategy->getJoinTableName($association, $class, $this->platform);
+        list($joinTargetEntitySQL, $filterSql) = $this->getFilterSql($mapping);
+
+        if ($filterSql) {
+            $conditions[] = $filterSql;
+        }
+
+        $sql = 'SELECT COUNT(*)'
+            . ' FROM ' . $joinTableName . ' t'
+            . $joinTargetEntitySQL
+            . ' WHERE ' . implode(' AND ', $conditions);
+
+        return $this->conn->fetchColumn($sql, $params);
     }
 
     /**
+     * {@inheritDoc}
+     */
+    public function slice(PersistentCollection $coll, $offset, $length = null)
+    {
+        $mapping   = $coll->getMapping();
+        $persister = $this->uow->getEntityPersister($mapping['targetEntity']);
+
+        return $persister->getManyToManyCollection($mapping, $coll->getOwner(), $offset, $length);
+    }
+    /**
      * {@inheritdoc}
-     *
-     * @throws \BadMethodCallException Not used for OneToManyPersister
      */
-    protected function getUpdateRowSQL(PersistentCollection $coll)
+    public function containsKey(PersistentCollection $coll, $key)
     {
-        throw new \BadMethodCallException("Insert Row SQL is not used for ManyToManyPersister");
+        list($quotedJoinTable, $whereClauses, $params) = $this->getJoinTableRestrictionsWithKey($coll, $key, true);
+
+        $sql = 'SELECT 1 FROM ' . $quotedJoinTable . ' WHERE ' . implode(' AND ', $whereClauses);
+
+        return (bool) $this->conn->fetchColumn($sql, $params);
     }
 
     /**
-     * {@inheritdoc}
+     * {@inheritDoc}
+     */
+    public function contains(PersistentCollection $coll, $element)
+    {
+        $entityState = $this->uow->getEntityState($element, UnitOfWork::STATE_NEW);
+
+        if ($entityState === UnitOfWork::STATE_NEW) {
```

Comment:
This code is duplicated in `contains` and `removeElement`: stash it into a private method